pkg/controllers/machine_controller.go

```diff
@@ -290,65 +298,133 @@ func (r *MachineReconciler) deleteMachine(ctx context.Context, log logr.Logger,
+// destroyDomain will do graceful destroy of domain
+func (r *MachineReconciler) destroyDomain(log logr.Logger, domain libvirt.Domain) error {
+	log.V(1).Info("Destroying domain")
+	// DomainDestroyFlags is blocking, so it can create performance issue in future
+	// During test with 26 emptyDisks function call spent max 1s.
+	// Performance issue will be in edge cases, if machine is correctly booted and it ignored shutdown command.
+	err := r.libvirt.DomainDestroyFlags(domain, libvirt.DomainDestroyGraceful)
+	if err != nil && !libvirt.IsNotFound(err) {
+		return err
 	}
 
-	log.V(1).Info("Destroying domain if exists")
-	if err := r.libvirt.DomainDestroy(domain); err != nil {
-		if !libvirtutils.IsErrorCode(err, libvirt.ErrNoDomain) {
-			return fmt.Errorf("error destroying domain: %w", err)
-		}
+	log.V(1).Info("Successfully destroyed domain")
+	return nil
+}
 
-		log.V(1).Info("Domain is already gone")
-		return nil
+func (r *MachineReconciler) updateDomainMetadata(log logr.Logger, domain libvirt.Domain, metadata *providermeta.LibvirtProviderMetadata) error {
+	log.V(1).Info("Updating domain metadata")
+	// Only 3 fields have flexible size.
+	// Range is between 245 - 291 bytes
+	// To prevent buffer reallocation, extra bytes are included in the initial size.
+	const xmlLength = 350
+	buf := bytes.NewBuffer(make([]byte, 0, xmlLength))
+	enc := xml.NewEncoder(buf)
+	err := metadata.MarshalXMLWithoutNamespace(enc)
+	if err != nil {
+		return err
 	}
-	log.V(1).Info("Successfully destroyed domain")
 
-	log.V(1).Info("Undefining domain if exists")
-	if err := r.libvirt.DomainUndefine(domain); err != nil {
-		if !libvirtutils.IsErrorCode(err, libvirt.ErrNoDomain) {
-			return fmt.Errorf("error undefining domain: %w", err)
-		}
+	return r.libvirt.DomainSetMetadata(domain, int32(libvirt.DomainMetadataElement), libvirt.OptString{buf.String()}, libvirt.OptString{"libvirtprovider"}, libvirt.OptString{"https://github.com/ironcore-dev/libvirt-provider"}, libvirt.DomainAffectCurrent)
+}
 
-		log.V(1).Info("Domain is already gone")
-		return nil
+func (r *MachineReconciler) cleanupAdditionalResources(ctx context.Context, log logr.Logger, machine *api.Machine) error {
+	log.V(1).Info("Removing volumes")
+	if err := r.deleteVolumes(ctx, log, machine); err != nil {
+		return fmt.Errorf("error removing machine disks: %w", err)
 	}
-	log.V(1).Info("Successfully undefined domain")
+	log.V(1).Info("Successfully removed machine disks")
+
+	log.V(1).Info("Removing network interfaces")
+	if err := r.deleteNetworkInterfaces(ctx, log, machine); err != nil {
+		return fmt.Errorf("error removing machine network interfaces: %w", err)
+	}
+	log.V(1).Info("Successfully removed network interfaces")
+
+	log.V(1).Info("Removing machine directory")
+	if err := os.RemoveAll(r.host.MachineDir(machine.ID)); err != nil {
+		return fmt.Errorf("error removing machine directory: %w", err)
+	}
+	log.V(1).Info("Successfully removed machine directory")
```

Comment:
We should avoid logging which printing "successful" in the message. Instead we should use e.g. "removing XYZ" when initiating an action and "removed XYZ" when the process is finished. The definition of what "successful" means is most of the time ambiguous.

Reply:
Thanks for your suggestion. It is now followed.